osu.Game.Tests/Visual/Gameplay/TestSceneBeatmapSkinFallbacks.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using NUnit.Framework;
+using osu.Framework.Allocation;
+using osu.Framework.Audio;
+using osu.Framework.Graphics;
+using osu.Framework.Lists;
+using osu.Framework.Testing;
+using osu.Framework.Timing;
+using osu.Framework.Utils;
+using osu.Game.Beatmaps;
+using osu.Game.Extensions;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Osu;
+using osu.Game.Rulesets.Osu.Skinning.Legacy;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Screens.Play.HUD;
+using osu.Game.Skinning;
+using osu.Game.Storyboards;
+using osu.Game.Tests.Beatmaps;
+
+namespace osu.Game.Tests.Visual.Gameplay
+{
+    public class TestSceneBeatmapSkinFallbacks : OsuPlayerTestScene
+    {
+        private ISkin currentBeatmapSkin;
+
+        [Resolved]
+        private SkinManager skinManager { get; set; }
+
+        [Cached]
+        private ScoreProcessor scoreProcessor = new ScoreProcessor();
+
+        [Cached(typeof(HealthProcessor))]
+        private HealthProcessor healthProcessor = new DrainingHealthProcessor(0);
+
+        protected override bool HasCustomSteps => true;
+
+        [Test]
+        public void TestEmptyDefaultBeatmapSkinFallsBack()
+        {
+            CreateSkinTest(DefaultLegacySkin.Info, () => new TestWorkingBeatmap(CreateBeatmap(new OsuRuleset().RulesetInfo)).Skin);
+            AssertComponentsFromExpectedSource(SkinnableTarget.MainHUDComponents, () => skinManager.CurrentSkin.Value);
+        }
+
+        protected void CreateSkinTest(SkinInfo gameCurrentSkin, Func<ISkin> beatmapSkin)
+        {
+            CreateTest(() =>
+            {
+                AddStep("setup skins", () =>
+                {
+                    skinManager.CurrentSkinInfo.Value = gameCurrentSkin;
+                    currentBeatmapSkin = beatmapSkin();
+                });
+            });
+        }
+
+        protected void AssertComponentsFromExpectedSource(SkinnableTarget target, Func<ISkin> expectedSource)
+        {
+            AddAssert($"{target} from {expectedSource.GetType().Name}", () =>
+            {
+                var expectedComponentsContainer = (SkinnableTargetComponentsContainer)expectedSource().GetDrawableComponent(new SkinnableTargetComponent(target));
+
+                Add(expectedComponentsContainer);
+                expectedComponentsContainer?.UpdateSubTree();
+                var expectedInfo = expectedComponentsContainer?.CreateSkinnableInfo();
+                Remove(expectedComponentsContainer);
+
+                var actualInfo = Player.ChildrenOfType<SkinnableTargetContainer>().First(s => s.Target == target)
+                                       .ChildrenOfType<SkinnableTargetComponentsContainer>().Single().CreateSkinnableInfo();
+
+                return almostEqual(actualInfo, expectedInfo, 2f);
```

Comment:
I've sneaked in a `2f` position tolerance as apparently, for some reason, the position of the accuracy counter in the legacy HUD components set may have a very slightly different calculated Y value.

I haven't deeply checked why this is the case yet, currently thinking there's some sort of weird offset between the expected target components container, and the actual target components container, will dive further into this in a bit.

Reply:
I've improved this, it actually just needed `0.001f` tolerance instead, not sure why it was failing for me before (probably something else I've mixed this up with)

I was initially going to put the equality logic in `SkinnableInfo` instead via `IEquatable`, but since this doesn't check for all properties to be *exactly* matching, I've just left it as-is.